Formula/b/bornagain.rb

```diff
@@ -0,0 +1,183 @@
+
+  license "GPL-3.0-or-later"
+
+  depends_on :macos
+  depends_on arch: [:arm64, :x86_64]
+
+  depends_on "cmake" => :build
+  depends_on "fftw"
+  depends_on "gsl"
+  depends_on "libcerf"
+  depends_on "libformfactor"
+  depends_on "libheinz"
+  depends_on "libtiff"
+  depends_on "python"
+  depends_on "qt"
+
+  resource "setuptools" do
+    url "https://files.pythonhosted.org/packages/a9/5a/0db4da3bc908df06e5efae42b44e75c81dd52716e10192ff36d0c1c8e379/setuptools-78.1.0.tar.gz"
+    sha256 "18fd474d4a82a5f83dac888df697af65afa82dec7323d09c3e37d1f14288da54"
+  end
+
+  resource "numpy" do
+    url "https://files.pythonhosted.org/packages/e1/78/31103410a57bc2c2b93a3597340a8119588571f6a4539067546cb9a0bfac/numpy-2.2.4.tar.gz"
+    sha256 "9ba03692a45d3eef66559efe1d1096c4b9b75c0986b5dff5530c378fb8331d4f"
+  end
+
+  resource "matplotlib" do
+    url "https://files.pythonhosted.org/packages/2f/08/b89867ecea2e305f408fbb417139a8dd941ecf7b23a2e02157c36da546f0/matplotlib-3.10.1.tar.gz"
+    sha256 "e8d2d0e3881b129268585bf4765ad3ee73a4591d77b9a18c214ac7e3a79fb2ba"
+  end
+
+  resource "fabio" do
+    url "https://files.pythonhosted.org/packages/ac/47/cd067e985b8a2476024b64373538c7e2b65b53415b39229e253d168d6d78/fabio-2024.9.0.tar.gz"
+    sha256 "f873df51f468531c11aae7e0cd88a14f221f4ef09431fbc5a6ca67b1ed47535b"
+  end
+
+  def install
+    build_dir = buildpath/"build/"
+    local_dir =  build_dir/"var/local/"
+    nproc = [Etc.nprocessors - 2, 1].max
+    cmake_exe = Formula["cmake"].opt_bin.realpath/"cmake"
+    python3_exe = "python3"
+    qt_cmake_dir = Formula["qt"].prefix/"lib/cmake/"
+    ff_cmake_dir = Formula["libformfactor"].prefix/"cmake/"
+    heinz_cmake_dir = Formula["libheinz"].prefix/"cmake/"
+    brew_bin_pth = (Pathname HOMEBREW_PREFIX/"bin/").realpath
+
+    # Build a Python virtual environment with the required packages
+    venv = virtualenv_create(libexec, "python3")
+    venv_root = venv.instance_variable_get(:@venv_root)
+    venv_py = venv_root/"bin/python3"
+    venv.pip_install resources
+
+    # Build BornAgain
+    ba_cmd = [cmake_exe, "-S", buildpath.to_s, "-B", build_dir.to_s,
+              *std_cmake_args, "-DBA_TESTS=OFF",
+              "-DCMAKE_PREFIX_PATH=#{qt_cmake_dir};#{ff_cmake_dir};#{heinz_cmake_dir};",
+              "-DBORNAGAIN_PYTHON=ON", "-DBA_PY_PACK=ON",
+              "-DPython3_EXECUTABLE=#{venv_py}"]
+
+    # CMake configuration step
+    system(*ba_cmd)
+
+    # CMake build step
+    system cmake_exe, "--build", build_dir.to_s,
+          "--config", "Release", "--parallel", nproc
+
+    system cmake_exe, "--build", build_dir.to_s,
+           "--config", "Release", "--target", "ba_wheel"
+
+    # CMake install step
+    system cmake_exe, "--install", build_dir.to_s, "--config", "Release"
+
+    # Install the BornAgain Python wheel in the virtual environment
+    ba_wheel = Dir.glob(build_dir/"py/wheel/*.whl").first
+    system venv_py, "-m", "pip", "install", ba_wheel
```

Comment:
Made a bunch of changes in this suggestion:
- The processes are set by brew, no need to adjust them
- build dir can just be relative
- Release config is default
- No need to use a full path for CMake
- No need to specify CMake dirs, brew will add them
- Use the same logic other formulae are using with venv
- Remove unneeded comments
```suggestion
    venv = virtualenv_create(buildpath/"venv", python3)
    venv.pip_install resources

    system "cmake", "-S", ".", "-B", "build",
              *std_cmake_args, "-DBA_TESTS=OFF",
              "-DBORNAGAIN_PYTHON=ON",
              "-DBA_PY_PACK=ON",
              "-DPython3_EXECUTABLE=#{buildpath}/venv/bin/python"

    system "cmake", "--build", "build"
    system "cmake", "--build", "build", "--target", "ba_wheel"
    system "cmake", "--install", "build"

    ba_wheel = Dir.glob(build_dir/"py/wheel/*.whl").first
    system #{buildpath}/venv/bin/python, "-m", "pip", "install", ba_wheel
```

Reply:
The virtual environment should _persist_ after the build to be used later by the BornAgain GUI to interoperate with the Python interpreter.

I was not sure how to achieve that.

Reply:
See Revision 4 of the formula, b2e8160f66ed782bb07a03c489632b67f46c15ad.

Reply:
Applied all the required changes in b785ff1dda4.